File: cc/stack/EvaluateReversePolishNotation.cc
#include <vector>
#include <string>
#include <stack>
#include <unordered_set>

#define TYPE_NUM 0
#define TYPE_OP  1

using namespace std;

/**
 * LeetCode 150. Evaluate Reverse Polish Notation
 * https://leetcode.com/problems/evaluate-reverse-polish-notation/
 */
class EvaluateReversePolishNotation {
public:
    int evalRPN(vector<string>& tokens) {
        init();
        stack<int> s;
        int n = tokens.size();
        for (int i = 0; i < n; i++) {
            switch (findType(tokens[i])) {
            case TYPE_NUM:
                s.push(stoi(tokens[i]));
                break;
            case TYPE_OP:
                doOperation(s, tokens[i]);
                break;
            default:
                break;
            }
        }
        return s.top();
    }
private:
    unordered_set<string> opSet;
    void init() {
        opSet.insert("+");
        opSet.insert("-");
        opSet.insert("*");
        opSet.insert("/");
    }
    void doOperation(stack<int>& s, string& op) {
        int a = s.top(); s.pop();
        int b = s.top(); s.pop();
        if (op == "+") {
            s.push(b + a);
        } else if (op == "-") {
            s.push(b - a);
        } else if (op == "*") {
            s.push(b * a);
        } else if (op == "/") {
            s.push(b / a);
        }
    }
    int findType(string& token) {
        if (opSet.find(token) != opSet.end()) {
            return TYPE_OP;
        }
        return TYPE_NUM;
    }
};

int main(int argc, char const *argv[]) {
    // ...
    return 0;
}

/* EOF */

File: typescript/stack/EvaluateReversePolishNotation.ts
/**
 * LeetCode 150. Evaluate Reverse Polish Notation
 * https://leetcode.com/problems/evaluate-reverse-polish-notation/
 */
function evalRPN(tokens: string[]): number {
    let opSet: Set<string> = new Set<string>();
    opSet.add("+");
    opSet.add("-");
    opSet.add("*");
    opSet.add("/");
    let s: Array<number> = new Array<number>();
    for (let i: number = 0; i < tokens.length; i++ ) {
        if (opSet.has(tokens[i])) { // Operator
            let a: number = s.pop();
            let b: number = s.pop();
            switch (tokens[i]) {
            case "+":
                s.push(b + a);
                break;
            case "-":
                s.push(b - a);
                break;
            case "*":
                s.push(b * a);
                break;
            case "/":
                s.push((b / a) | 0);
                break;
            default:
                break;
            }
        } else { // Number
            s.push(parseInt(tokens[i]));
        }
    }
    return s[s.length - 1];
};

// EOF
